test/puppetlabs/puppetdb/testutils/db.clj

```diff
@@ -128,51 +135,63 @@
         (jdbc/do-commands-outside-txn
          "create extension if not exists pg_trgm"
          "create extension if not exists pgcrypto"))
-      (let [cfg (db-user-config template-name)]
+      (let [cfg (routine-db-config template-name)]
         (jdbc/with-db-connection cfg
-          (migrate!)))
+          (initialize-schema)))
       (reset! template-created true))))
 
+(defn- require-suitable-pg-arrangement
+  [{:keys [user migrator-username subname] :as config}]
+  (jdbc/with-db-connection config
+    (let [db (subname->validated-db-name subname)
+          priv? jdbc/has-database-privilege?
+          migrator migrator-username]
+      (when (priv? "public" db "connect")
+        (throw (Exception. "public connect")))
+      (when-not (priv? user db "connect")
+        (throw (Exception. "no user connnect")))
+      (when (priv? user db "connect with grant option")
+        (throw (Exception. "user has connect grant")))
+      (when-not (priv? migrator db "connect with grant option")
+        (throw (Exception. "no migrator connect grant")))
+      (when-not (jdbc/has-role? migrator user "member")
+        (throw (Exception. "migrator not member of user"))))))
+
 (def ^:private test-db-counter (atom 0))
 
 (defn create-temp-db []
   "Creates a temporary test database.  Prefer with-test-db, etc."
   (ensure-pdb-db-templates-exist)
   (let [n (swap! test-db-counter inc)
-        db-name (if-not pdb-test-id
-                  (str "pdb_test_" n)
-                  (str "pdb_test_" pdb-test-id "_" n))]
-    (assert (valid-sql-id? db-name))
+        db (if-not pdb-test-id
+             (str "pdb_test_" n)
+             (str "pdb_test_" pdb-test-id "_" n))
+        db-q (jdbc/double-quote db)
+        config (routine-db-config db)
+        user (get-in test-env [:user :name])
+        user-q (jdbc/double-quote user)
+        migrator (get-in test-env [:migrator :name])
+        migrator-q (jdbc/double-quote migrator)]
     (jdbc/with-db-connection (db-admin-config)
       (jdbc/do-commands-outside-txn
-       (format "drop database if exists %s" db-name)
-       (format "create database %s template %s" db-name template-name)))
-    (db-user-config db-name)))
+       (format "drop database if exists %s" db-q)
+       (format "create database %s template %s" db-q template-name)
+       ;; Needed by migration coordination
+       (format "revoke connect on database %s from public" db-q)
+       (format "grant connect on database %s to %s with grant option" db-q migrator-q)
+       (format "set role %s" migrator-q)
+       (format "grant connect on database %s to %s" db-q user-q)
+       "commit"))
```

Comment:
I think the extra "commit" statement here is causing the _WARNING:  there is no transaction in progress_ pg logs we've been seeing. I think we may need a "begin" at the top of the statements here if we want them grouped into a transaction.

Reply:
Ahh, right - left over from some diagnostics.  I thought I'd removed that.